### PR TITLE
Protected Positions Table

### PR DESCRIPTION
**Pull request by pingustar**
#249 #345

### CODE REVIEW

**Comment by RanCohenn** [comment]
- [ ] Pool tokens should not shows when there are no pool tokens
- [ ] Table loading instead of showing empty state when positions finished loading and are empty
- [ ] My stake should show loading state when table is loading
- [ ] Rewards summary should show loading state when loading
- [ ] Connect withdraw to protected position

**Review by zavelevsky — APPROVED** [review]
LGTM

File: src/elements/earn/portfolio/withdrawLiquidity/WithdrawLiquidityWidget.tsx [hunk 1]
```diff
@@ -1,80 +1,151 @@
...
+
+export const WithdrawLiquidityWidget = ({
+  protectedPosition,
+  isModalOpen,
+  setIsModalOpen,
+}: Props) => {
+  const { chainId } = useWeb3React();
+  const { positionId, reserveToken, currentCoveragePercent, pool } =
+    protectedPosition;
+  const { tknAmount } = protectedPosition.claimableAmount;
```
[the remark this answers is not in the document]
Comment (RanCohenn) [reply src/elements/earn/portfolio/withdrawLiquidity/WithdrawLiquidityWidget.tsx:36]: I know you can do that spread but in this case Its not going to work
`protectedPosition.claimableAmount.tknAmount`
So you cant do it like that (The code will need to be changed to `tknAmount.tknAmount`)

File: src/elements/earn/portfolio/withdrawLiquidity/WithdrawLiquidityWidget.tsx [hunk 2]
```diff
@@ -1,80 +1,151 @@
...
+  const protectionNotReached = currentCoveragePercent !== 1;
   const multiplierWillReset = true;
 
   const outputBreakdown = [
     { color: 'blue-4', decPercent: 0.75, label: 'ETH' },
     { color: 'primary', decPercent: 0.25, label: 'BNT' },
   ];
 
+  useAsyncEffect(async (isMounted) => {
+    if (isMounted())
```
[the remark this answers is not in the document]
Comment (RanCohenn) [reply src/elements/earn/portfolio/withdrawLiquidity/WithdrawLiquidityWidget.tsx:56]: So useRef current state doesn't count?

**Comment by ashachaf** [comment]
- [ ] main table loading -> please set it to "load" 3 placeholder rows
current: 
image
- [ ] on mobile, please add some padding and make sure the text is not cut off
image
- [ ] when connecting wallet, i get these console errors: 
image
- [ ] on the portfolio page -> connect wallet -> logout (click the CTA on the top right) -> i end up seeing a blank page. expected to see the portfolio empty state.
image

- [ ] withdraw protection popup, output breakdown
IF the user did not indicate any amount -> do not show the breakdown
IF the user input an amount -> check the breakdown -> IF some amount is returned in BNT, show the breakdown
- [ ] pool table, expand group is missing indication
current: 
image
updated: 
image
mockup:
https://www.figma.com/file/fwADI9wqDrRAdlMX8EddCw/Bancor-v3?node-id=8213%3A218341
- [ ] withdraw position -> IF user confirm the transaction on metamask and initiated the process -> close the withdraw popup
- [ ] once a withdraw is confirmed on the blockchain -> force update the data in the table. if needed, remove the position that was removed or update the available balance. also, this should immediately affect the rewards multiplier we show.
- [ ] approve notification
- [ ] withdraw notification
- [ ] open withdraw BNT -> click to withdraw -> see approve -> close  -> scroll get stuck
- [ ] withdraw popup -> when input amount =0, disable the CTA. update text only when input amount = 0: current: "withdraw", updated: "Enter amount"

**Comment by ashachaf** [comment]
- [ ] incorrect protected positions - i was able to view protected position of a wallet that is not the current one connected to the app. edge case. 
flow:
- connect wallet A (wallet with positions)
- switch to wallet B (wallet with no positions
- expected - see the information of wallet B.
- actual - see the information of wallet A.
wallet A: 
image
switch to wallet B (open metamask and switch wallet):
image

- [ ] connecting wallet with no protected positions -> infinite load
image
no visible console error

- [ ] withdraw position modal
when input amount > available claim amount -> display error message "Token balance is currently insufficient"
reference from the trade widget:
image

- [ ] withdraw position modal - align design with mockup
https://www.figma.com/file/fwADI9wqDrRAdlMX8EddCw/Bancor-v3?node-id=8149%3A211892

- [ ] update position data quicker
flow:
- open a position
- withdraw 50%
- process
- confirm on metamask
- wait for successful completed notification
- click on the same position
- withdraw 100%
- expected - balance should be after the last withdraw. 
- bonus issue #1- if you proceed, there is no metamask error for insufficient balance
- bonuse issue #2 - failed transaction notification shows up twice
image
with big error message in console:
```
positions.ts:526 Error: transaction failed (transactionHash="0xaffbb95f1b956693efaf6f3e9d69302eb508286661943138f8e5492d65eccdd0", transaction={"hash":"0xaffbb95f1b956693efaf6f3e9d69302eb508286661943138f8e5492d65eccdd0","type":2,"accessList":null,"blockHash":null,"blockNumber":null,"transactionIndex":null,"confirmations":0,"from":"0x5f7a009664B771E889751f4FD721aDc439033ECD","gasPrice":{"type":"BigNumber","hex":"0x59682f0e"},"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"maxFeePerGas":{"type":"BigNumber","hex":"0x59682f0e"},"gasLimit":{"type":"BigNumber","hex":"0x0a5458"},"to":"0x99427BC8a0336EF885bFC0B4316e0003BE183D18","value":{"type":"BigNumber","hex":"0x00"},"nonce":1109,"data":"0x782ed90c000000000000000000000000000000000000000000000000000000000000002d00000000000000000000000000000000000000000000000000000000000f4240","r":"0x77f316b58b9400af33f20dee7d81aaa2940f9f14c5e5d180f301a19fab50ef2f","s":"0x2fe0e1043680dbfe0953879c67454b6e1e2e9dd92091d06127a77f232efa8ef3","v":1,"creates":null,"chainId":0}, receipt={"to":"0x99427BC8a0336EF885bFC0B4316e0003BE183D18","from":"0x5f7a009664B771E889751f4FD721aDc439033ECD","contractAddress":null,"transactionIndex":4,"gasUsed":{"type":"BigNumber","hex":"0x0a53c1"},"logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","blockHash":"0x2ecb3bb084eec2d907157ffcdcd5e8d601794952692853ac542d74c1c9124552","transactionHash":"0xaffbb95f1b956693efaf6f3e9d69302eb508286661943138f8e5492d65eccdd0","logs":[],"blockNumber":11391609,"confirmations":2,"cumulativeGasUsed":{"type":"BigNumber","hex":"0x1d74c8"},"effectiveGasPrice":{"type":"BigNumber","hex":"0x59682f0a"},"status":0,"type":2,"byzantium":true}, code=CALL_EXCEPTION, version=providers/5.4.5)
```

- [ ] trying to withdraw BNT (from the eth/bnt pool on ropsten) -> withdraw modal CTA is always disabled
image

**Comment by zavelevsky** [comment]
- [ ] when switching networks we should indicate loading and also hide the logos. Now they show this state for a few seconds
 
image

**Comment by zavelevsky** [comment]
opened issues for the unchecked items, let's stop corresponding over this PR

**Pull request merged**